### PR TITLE
Fix PollingFileSystemWatcher unit tests

### LINKED ISSUE #420: Intermittent test failure in System.IO.FileSystem.Watcher.Polling.Test
```
19:03:57   Discovering: System.Text.Formatting.Tests
19:03:57   Discovered:  System.Collections.Tests
19:03:57   Discovered:  System.IO.FileSystem.Watcher.Polling.Tests
19:03:57   Discovered:  System.Text.Formatting.Tests
19:03:57   Starting:    System.Collections.Tests
19:03:58   Starting:    System.IO.FileSystem.Watcher.Polling.Tests
19:03:58   Starting:    System.Text.Formatting.Tests
19:03:58   InvariantFormatStructClr : Elapsed 722ms
19:03:59   
19:03:59   Unhandled Exception: System.ObjectDisposedException: Cannot access a disposed object.
19:03:59      at System.Threading.TimerQueueTimer.Change(UInt32 dueTime, UInt32 period)
19:03:59      at System.IO.FileSystem.PollingWatcher.TimerHandler(Object context)
19:03:59      at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
19:03:59      at System.Threading.TimerQueueTimer.CallCallback()
19:03:59      at System.Threading.TimerQueueTimer.Fire()
19:03:59      at System.Threading.TimerQueue.FireNextTimers()
19:03:59   InvariantFormatIntHexClr : Elapsed 668ms
19:04:00 d:\j\workspace\dotnet_corefxlab\Debug_prtest\packages\Microsoft.DotNet.BuildTools.1.0.25-prerelease-00107\lib\tests.targets(129,5): warning MSB3073: The command "CoreRun.exe xunit.console.netcore.exe System.IO.FileSystem.Watcher.Polling.Tests.dll  -xml testResults.xml -notrait category=nonwindowstests  -notrait category=OuterLoop -notrait category=failing " exited with code -532462766. [d:\j\workspace\dotnet_corefxlab\Debug_prtest\src\System.IO.FileSystem.Watcher.Polling\tests\System.IO.FileSystem.Watcher.Polling.Tests.csproj]
19:04:00 d:\j\workspace\dotnet_corefxlab\Debug_prtest\packages\Microsoft.DotNet.BuildTools.1.0.25-prerelease-00107\lib\tests.targets(136,5): error : One or more tests failed while running tests from 'System.IO.FileSystem.Watcher.Polling.Tests' please check log for details! [d:\j\workspace\dotnet_corefxlab\Debug_prtest\src\System.IO.FileSystem.Watcher.Polling\tests\System.IO.FileSystem.Watcher.Polling.Tests.csproj]
```

http://dotnet-ci.cloudapp.net/job/dotnet_corefxlab/job/Debug_prtest/34/console

### PR DESCRIPTION
**Pull request by MisinformedDNA**
Fixes #420

### CODE REVIEW

**Comment by MisinformedDNA** [comment]
The Ubuntu error is 

> 19:48:23 The active test run was aborted. Reason: Unhandled Exception: System.NullReferenceException: Object reference not set to an instance of an object.
> 19:48:23    at System.IO.FileSystem.PollingFileSystemWatcher.TimerHandler(Object context) in /mnt/j/workspace/dotnet_corefxlab/master/ubuntu16.04_debug_prtest/src/System.IO.FileSystem.Watcher.Polling/System/IO/PollingFileSystemWatcher.cs:line 154
> 19:48:23    at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
> 19:48:23 --- End of stack trace from previous location where exception was thrown ---
> 19:48:23    at System.Threading.TimerQueueTimer.CallCallback()
> 19:48:23    at System.Threading.TimerQueueTimer.Fire()
> 19:48:23    at System.Threading.TimerQueue.FireNextTimers()

The error is on [line 154](https://github.com/MisinformedDNA/corefxlab/blob/fcfb5ca136116be080db8534d5ec532fb481cca5/src/System.IO.FileSystem.Watcher.Polling/System/IO/PollingFileSystemWatcher.cs#L154).

Ideas?

**Comment by MisinformedDNA** [comment]
I was already waiting for the event to complete, up to one second, with https://github.com/dotnet/corefxlab/pull/2253/files#diff-a05cc2666ae2db170969b1cf419f5892R29. `Timer.Dispose(WaitHandle)` seems to be a different way to reach the same goal. I still don't see where my code could be going wrong. Whether a callback was called after Dispose or something else, I would expect an `ObjectDisposedException`, not a `NullReferenceException`.

In any case, what's the best way to run the Ubuntu innerloop on my machine? I'll play around with it and see if I can't figure out what's happening.

**Comment by JeremyKuhne** [comment]
> I still don't see where my code could be going wrong.

Accessing the field on the callback worries me. Is it possible for the timer to fire before it is assigned to the field? With a brief look, it seems like it it could happen. You could call https://msdn.microsoft.com/en-us/library/ms149618(v=vs.110).aspx first, then start the first period.

I'd still wait in the watcher's Dispose() for the timer events to clear. Seems like bad mojo to start the timer again on a trailing event. I'd also consider making the timer itself the state object as described in the constructor link above.

**Comment by MisinformedDNA** [comment]
`_timer` is first accessed in the `Start` method. It is impossible for any events to happen before this assignment because it is the `TimerCallback`, `TimerHandler`, that is called by the timer to calculate and raise any changes. Calling `TimerCallback` doesn't help because we are essentially doing that in the `Start` method.

Starting the timer again at the end of each callback is how we ensure that there is only one callback running at a time. How else would we do it?

**Comment by JeremyKuhne** [comment]
> _timer is first accessed in the Start method. It is impossible for any events to happen before this assignment because it is the TimerCallback, TimerHandler, that is called by the timer to calculate and raise any changes. Calling TimerCallback doesn't help because we are essentially doing that in the Start method.

`_timer` is set in `Start()`. Until the constructor for `Timer()` has finished `_timer` will be null. In the constructor of `Timer()` the timer starts (i.e, while `_timer` is still null). There is one shared timer for the VM that fires on unspecified threadpool threads. I can totally see this sometimes causing the callback to fire _before_ `_timer` is fully assigned. It may be even more likely on some platforms.

Bottom line- don't start the `_timer` before the instance is assigned. That requires two steps. If you're doing two steps anyway, you could consider passing the instance as the state rather than null.

> Starting the timer again at the end of each callback is how we ensure that there is only one callback running at a time. How else would we do it?

That is the right way to do it. I'm saying don't let `PollingFileSystemWatcher.Dispose()` return before the timer has flushed it's last event. Or possibly provide an overload that takes a WaitHandle and passes it through to the timer.

**Comment by JeremyKuhne** [comment]
> I'm saying don't let PollingFileSystemWatcher.Dispose() return before the timer has flushed it's last event. Or possibly provide an overload that takes a WaitHandle and passes it through to the timer.

I presume that setting the next interval on the disposed timer is a no-op. Should take a look and/or consider keeping track of disposed state.

**Comment by MisinformedDNA** [comment]
Thanks! That is helpful.

**Comment by ahsonkhan** [comment]
@MisinformedDNA, can you please disable the tests, for the time being, and add a comment in source to link to #420? This way, other PRs remain unblocked by it.

**Comment by ahsonkhan** [comment]
> If we can get https://github.com/dotnet/corefxlab/pull/2258 merged, I can just fix these tests.

Ah yes, continue :)

**Comment by ahsonkhan** [comment]
@dotnet-bot test this please

**Comment by JeremyKuhne** [comment]
> Of course, I'm not seeing any on my box. Is there something in the output that I am missing? See any obvious mistakes?

Have you tried building/testing release?

**Comment by JeremyKuhne** [comment]
There is a specific note about calling `Dispose` when the `dueTime == 0` in the [docs]( https://docs.microsoft.com/en-us/dotnet/api/system.threading.timer.dispose?view=netcore-2.1#System_Threading_Timer_Dispose_System_Threading_WaitHandle_).

**Comment by MisinformedDNA** [comment]
All passed!

**Comment by MisinformedDNA** [comment]
Can we merge this?
